src/middlewared/middlewared/plugins/jbof/redfish/client.py

```diff
@@ -0,0 +1,300 @@
+
+    def __init__(self,
+                 base_url,
+                 username=None,
+                 password=None,
+                 authtype=AuthMethod.BASIC,
+                 default_prefix=REDFISH_ROOT_PATH,
+                 verify=False,
+                 timeout=DEFAULT_REDFISH_TIMEOUT_SECS):
+        self.base_url = base_url.rstrip('/')
+        self.username = username
+        self.password = password
+        self.authtype = self.authtype_to_enum(authtype)
+        self.prefix = default_prefix
+        self.verify = verify
+
+        self.auth = None
+        self.auth_token = None
+        self.session_key = None
+        self.authorization_key = None
+        self.session_location = None
+        self.timeout = timeout
+        self.cache = {}
+
+        self.root = self.get_root_object()
+        try:
+            self.login_url = self.root['Links']['Sessions']['@odata.id']
+        except KeyError:
+            self.login_url = '/redfish/v1/SessionService/Sessions'
+
+        if username and password:
+            self.login()
+
+    @property
+    def uuid(self):
+        return self.root['UUID']
+
+    @property
+    def product(self):
+        return self.root['Product']
+
+    def _members(self, data):
+        result = {}
+        for manager in data['Members']:
+            uri = manager[ODATA_ID]
+            result[uri.split('/')[-1]] = uri
+        return result
+
+    def _cached_fetch(self, cache_key, uri, use_cached=True):
+        if use_cached and cache_key in self.cache:
+            return self.cache[cache_key]
+
+        r = self.get(uri)
+        if r.ok:
+            self.cache[cache_key] = self._members(r.json())
+            return self.cache[cache_key]
+
+    def managers(self, use_cached=True):
+        return self._cached_fetch('managers', '/Managers', use_cached)
+
+    def mgmt_ethernet_interfaces(self, iom, use_cached=True):
+        uri = self.managers()[iom] + '/EthernetInterfaces'
+        return self._cached_fetch(f'{iom}/mgmt_ethernet_interfaces', uri, use_cached)
+
+    def network_device_functions(self, iom, use_cached=True):
+        return self._cached_fetch(f'{iom}/network_device_functions', f'/Chassis/{iom}/NetworkAdapters/1/NetworkDeviceFunctions', use_cached)
+
+    def fabric_ethernet_interfaces(self, use_cached=True):
+        result = []
+        for iom in self.managers(use_cached):
+            for ndfuri in self.network_device_functions(iom, use_cached).values():
+                result.append(f'{ndfuri}/EthernetInterfaces/1')
+        result.sort()
+        return result
+
+    def get_uri(self, uri, use_cached=True):
+        if use_cached and uri in self.cache:
+            return self.cache[uri]
+
+        r = self.get(uri)
+        if r.ok:
+            self.cache[uri] = r.json()
+            return self.cache[uri]
+
+    def get_root_object(self):
+        try:
+            return self.get(self.prefix).json()
+        except Exception as excp:
+            raise excp
```

Comment:
Did we want to do something with this exception?

Reply:
Logged